model/models/llama/model.go

```diff
@@ -86,13 +86,7 @@ func (sa *SelfAttention) Forward(ctx ml.Context, hiddenState, positionIDs ml.Ten
 	k = k.Permute(ctx, 0, 2, 1, 3).Contiguous(ctx)
 	v = v.Permute(ctx, 1, 2, 0, 3).Contiguous(ctx)
 
-	kq := k.MulmatFullPrec(ctx, q)
-	kq = kq.Scale(ctx, 1.0/math.Sqrt(float64(headDim)))
-	kq = kq.Add(ctx, mask)
-	kq = kq.Softmax(ctx)
-
-	kqv := v.Mulmat(ctx, kq)
-	kqv = kqv.Permute(ctx, 0, 2, 1, 3).Contiguous(ctx)
+	kqv := nn.Attention(ctx, q, k, v, mask, 1.0/math.Sqrt(float64(headDim)))
```

Comment:
It feels less clear to me how the scaling factor is set now and how its used, maybe we should move it to a constant somewhere in the model definition

Reply:
Can you elaborate a little on why this is less clear now? It's the same calculation of the scaling factor as before, just passed as a parameter to Attention rather than explicit call to Scale.

Reply:
I find it easier to eye-ball when looking at the model forward pass implementation when its directly associated with scaling:
```go
kq = kq.Scale(ctx, 1.0/math.Sqrt(float64(headDim)))
```
The fact it is being passed to scale make it apparent that this is probably the scaling factor.

```go
kqv := nn.Attention(ctx, q, k, v, mask, 1.0/math.Sqrt(float64(headDim)))
```
Now I can tell what this math operation is while reading the code, I have to follow through and see what it is actually used for in the attention implementation.

```go
var scaleFactor = 1.0/math.Sqrt(float64(headDim))
kqv := nn.Attention(ctx, q, k, v, mask, scaleFactor)
```
I personally like something like this, so I can understand what is going on without leaving the current file.

Reply:
That makes sense - I can pull it out as a named variable.